### PR TITLE
Create notifications config resources

### PR DESCRIPTION
**Pull request by jaideepr97**
**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`

> /kind bug
> /kind chore
> /kind cleanup
> /kind failing-test
 /kind enhancement
> /kind documentation
> /kind code-refactoring

**What does this PR do / why we need it**:

**NOTE**: This PR builds on top of https://github.com/argoproj-labs/argocd-operator/pull/635 and should only be merged after that one has been merged. This PR will be updated with the same review changes from https://github.com/argoproj-labs/argocd-operator/pull/635 after it has been merged. Review comments need not be repeated
 
This is 2/2 PRs for adding support for spinning up the notifications-controller within argocd-operator. This PR enables the operator to create/delete the `argocd-notifications-cm` configmap and `argcocd-notifications-secret` based on whether notifications has been enabled or not. It also adds a set of default template and triggers configurations to the cm out of the box, based on what is present in https://github.com/argoproj/argo-cd/blob/master/notifications_catalog/install.yaml

**Have you updated the necessary documentation?**

* [x] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:
https://github.com/argoproj-labs/argocd-operator/issues/655

**How to test changes / Special notes to the reviewer**:
1. Start the operator locally and spin up an empty Argo CD instance
2. Notice that out of the box there are no configmaps or secrets created for notifications
3. Set `.spec.notifications.enabled` to true
4. Verify that `argocd-notifications-cm` and `argocd-notifications-secret` are created 
5. Verify that `argocd-notifications-cm` contains a set of default templates and triggers out of the box
6. Disable notifications by setting `.spec.notifications.enabled` to false
7. Verify that the cm and secret are properly cleaned up

### CODE REVIEW

**Comment by ishitasequeira** [comment]
@jaideepr97 , this PR seems to contain changes from your earlier PR #635. Is this intentional? (Especially as this PR is raised on top of the other PR)

**Comment by jaideepr97** [comment]
@ishitasequeira yes
This is built on top of the other one 
Split them up since the stories were also split up that way, And it can only be tested all together 
This should go in only after the first one does

**Comment by ishitasequeira** [comment]
I understand, but do we need same changes in both the PRs? Or are you planning on updating this PR once the other one goes in?

File: controllers/argocd/notifications.go [hunk 1]
```diff
@@ -0,0 +1,1081 @@
...
+	if err := r.reconcileNotificationsRoleBinding(cr, role, sa); err != nil {
+		return err
+	}
+
+	log.Info("reconciling notifications deployment")
+	if err := r.reconcileNotificationsDeployment(cr, sa); err != nil {
+		return err
+	}
+
+	log.Info("reconciling notifications configmap")
```
[the remark this answers is not in the document]
Comment (jaideepr97) [reply controllers/argocd/notifications.go:46]: It does, thanks! I'll update this along with the other changes that have gone into https://github.com/argoproj-labs/argocd-operator/pull/635 once that has been merged

[the remark this answers is not in the document]
Comment (jaideepr97) [reply controllers/argocd/notifications.go:41]: done

File: controllers/argocd/notifications.go [hunk 2]
```diff
@@ -413,6 +524,547 @@ func getNotificationsResources(cr *argoprojv1a1.ArgoCD) corev1.ResourceRequireme
 	return resources
 }
 
+// getDefaultNotificationsConfig returns a map that contains default triggers and template configurations for argocd-notifications-cm
+func getDefaultNotificationsConfig() map[string]string {
```
Comment (ishitasequeira) [inline controllers/argocd/notifications.go:528]: Just wondering if shifting this to a different file would help keep file easily readable for future. The file just looks too big to me especially as we are planning to refactor the codebase into smaller chunks.

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
LGTM! Tested the workflow and everything works as expected. Just some minor comments.

File: controllers/argocd/notifications.go [hunk 3]
```diff
@@ -393,6 +413,97 @@ func (r *ReconcileArgoCD) reconcileNotificationsDeployment(cr *argoprojv1a1.Argo
...
+	if !cr.Spec.Notifications.Enabled {
+		return nil
+	}
+
+	// secret doesn't exist but should, so it should be created
+	if err := controllerutil.SetControllerReference(cr, desiredSecret, r.Scheme); err != nil {
+		return err
+	}
+
+	log.Info(fmt.Sprintf("Creating configmap %s", desiredSecret.Name))
```
[the remark this answers is not in the document]
Comment (jaideepr97) [reply controllers/argocd/notifications.go:498]: ah, yeah
thanks for catching!

File: controllers/argocd/notifications.go (hunk 2 above)
Comment (jaideepr97) [reply controllers/argocd/notifications.go:528]: Yeah, this should go into a separate `notifications_util.go` or something 
but I thought that would be better suited to the refactoring task - I didn't want to conflate the 2 efforts in a single PR

Comment (jaideepr97) [reply controllers/argocd/notifications.go]: done

**Review by ishitasequeira — APPROVED** [review]
LGTM!

**Comment by jaideepr97** [comment]
Thanks for the feedback @ishitasequeira !

**Review by jopit — APPROVED** [review]
LGTM

**Pull request merged**